lib/client/session.go

```diff
@@ -543,12 +544,28 @@ func (ns *NodeSession) pipeInOut(shell io.ReadWriteCloser) {
 	go func() {
 		defer ns.closer.Close()
 		buf := make([]byte, 128)
+
+		stdin := ns.stdin
+		if ns.isTerminalAttached() {
+			stdin = escape.NewReader(stdin, ns.stderr, func(err error) {
+				switch err {
+				case escape.ErrDisconnect:
+					fmt.Fprintf(ns.stderr, "\r\n%v\r\n", err)
+				case escape.ErrTooMuchBufferedData:
+					fmt.Fprintf(ns.stderr, "\r\nerror: %v\r\nremote peer may unreachable, check your connectivity\r\n", trace.Wrap(err))
+				default:
+					fmt.Fprintf(ns.stderr, "\r\nerror: %v\r\n", trace.Wrap(err))
+				}
+				ns.closer.Close()
```

Comment:
It looks like this closer is always going to be called as it's outside the `switch` block? Not sure if that's intentional or if I've misunderstood its function - it seems like it this be overenthusiastically closing the connection once attached to `stdin` even if there was no error received.

Reply:
this callback is only called with non-nil errors.
it's meant as a signal for the parent to prematurely terminate the connection.
it won't be called when the connection exits cleanly.